lib/galaxy/config/__init__.py

```diff
@@ -262,21 +262,20 @@ def _process_config(self, kwargs):
             self.thread_local_log = threading.local()
 
         # Install database related configuration (if different).
-        self.install_database_connection = kwargs.get("install_database_connection", None)
         self.install_database_engine_options = get_database_engine_options(kwargs, model_prefix="install_")
 
         # Where dataset files are stored
-        self.file_path = resolve_path(kwargs.get("file_path", "files"), self.data_dir)
+        self.file_path = resolve_path(self.file_path, self.data_dir)
         # new_file_path and legacy_home_dir can be overridden per destination in job_conf.
-        self.new_file_path = resolve_path(kwargs.get("new_file_path", "tmp"), self.data_dir)
+        self.new_file_path = resolve_path(self.new_file_path, self.data_dir)
```

Comment:
I think there is a serious problem, introduced in #921 (and so present also in the `release_19.09` branch), with resolving relative paths with respect to `self.data_dir` or anything that's not `self.root` .
The issue, which I mentioned already in https://github.com/galaxyproject/galaxy/pull/8490#discussion_r316313454 , is that #921 changed how these settings are interpreted if the admin used a relative path, from being relative to `self.root` to being relative to e.g. `self.data_dir`.

Something like https://github.com/galaxyproject/galaxy/pull/8490/files#diff-46f86ea0d82b84642e4218279ef61442R582 can be used to fix the issue for 19.09 , not sure how to integrate it with the changes in `dev` though.

Ping @natefoo @ic4f @jmchilton

Reply:
The referenced code doesn't change the way `self.new_file_path` is resolved. The 2 lines are equivalent: 
1. 'tmp' is assigned to `self.new_file_path `as the default:
https://github.com/galaxyproject/galaxy/pull/8639/files#diff-a73e218444bd4a6820a525616a23f40dR186
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L198
2. It is overwritten by the value in kwargs:
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L199
3. And all is resolved to `self.data_dir` (which, indeed *may be relative* if a value is provided here:
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L129 - but that's a separate issue, I think).

Unless I'm mistaken, that's essentially the same fix as in [the edit you pointed to](https://github.com/galaxyproject/galaxy/pull/8490/files#diff-46f86ea0d82b84642e4218279ef61442R582)

The purpose of this refactoring, along with many other similar edits is to simplify all that code before making more significant modifications.

I think we should open a separate issue for `data_dir` resolution (and related stuff). @natefoo and I have been discussing this (path resolution in config) over the past several days. I'm working on this at the moment (and happy to discuss, of course!)

Reply:
We're going to try to address that with something like what I've done [here](https://github.com/natefoo/galaxy/commit/3a3279b497f790b113d6cf2721b64b148d0087fb).

Reply:
@ic4f As said above, the issue originates in #921, not here. But this change may make it trickier to "restore" the original defaults.

> Unless I'm mistaken, that's essentially the same fix as in the edit you pointed to

That's not the case, the fix I linked makes the default value resolve w.r.t. `self.data_dir`, while a user-defined value would be resolved w.r.t. to `self.root`, without breaking the old behaviour.

@natefoo I've just had a look at your proposed solution, I think I have ambivalent feelings:
- is it over-complicated?
- is it better to warn admins to update their config file, and have to update our documentation/training, or would it better to keep the old behaviour for user-specified values?

Reply:
@nsoranzo I see what you mean. However, the 2 lines you refer to in this commit are identical in terms of what they do. I don't think this edit makes it tricker to fix the data_dir resolution issue; quite the contrary, it removes redundant code: kwargs have already been read, and "tmp" has been already used as the default. Without this edit, we are doing those things twice. (and not just new_file_path). 

Resolving a path w.r.t one absolute dir or another can be automated, based on, for example, the value of a `path_resolves_to` attribute specified in the schema. That, and some logic similar to what is done in BaseAppConfiguration (in `_set_config_base`). 

But again, this particular commit, I think, doesn't preclude any of this.

Reply:
@nsoranzo @natefoo : may I mark this as resolved? There are several commits depending on this one. Besides, there's one coming up on path resolution (which is where we can discuss how to account for old configs that may have "database/" and "config/" prepended to user-defined values, if that commit does not offer a satisfactory solution).
If not, that's perfectly fine - I'm happy to discuss and revise/resubmit, etc.

Reply:
Marking as resolved. I guess.